sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.json
+
+import org.apache.spark.sql.catalyst.{InternalRow, StructFilters}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.types.StructType
+
+/**
+ * The class provides API for applying pushed down source filters to rows with
+ * struct schema parsed from JSON records. It assumes that:
+ *   1. reset() is called before any skipRow() calls for new row.
+ *   2. skipRow() can be called for any valid index of the struct fields,
+ *      and in any order.
+ *
+ * @param pushedFilters The pushed down source filters. The filters should refer to
+ *                      the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(pushedFilters: Seq[sources.Filter], schema: StructType)
```

Comment:
can we have a big picture to explain how it works? e.g.
1. the json parser reads out one entire json record
2. iterate the json fields, and run the corresponding predicate for each field
3. if the json record can't be skipped, convert to internal row.

And also explain how to define the predicate for each field. IIUC:
1. for the first field, run the predicate that only refer to the first field
2. for the second field, run the predicate that refers to both the first and second field
3. ....

Reply:
> 1. the json parser reads out one entire json record

No-no, Jackson parser shouldn't read entire record. Of course, it does buffering and looks ahead but it reads JSON fields in stream fashion while we pull JSON fields via iterator-like API. 

> 2. iterate the json fields, and run the corresponding predicate for each field
> 3. if the json record can't be skipped, convert to internal row.

Actually, we convert a value first of all and set it to the internal row. After that, we invoke predicates for partially initialized internal row. 

ok. Let me write the high level view somewhere. Don't think `JsonFilters` is the right place, probably `StructFilters`, because the logic is applicable to `CSVFilters` as well.

> And also explain how to define the predicate for each field. IIUC:

Are the comments with examples https://github.com/apache/spark/pull/27366/files#diff-467cfe570f88d176aed53f30c3766356R57-R104 not enough?

Reply:
I added a comment for JsonFilters